contrib/raftexample/raftchan.go

```diff
@@ -0,0 +1,240 @@
+
+		w, err := wal.Create(waldir, nil)
+		if err != nil {
+			log.Fatalf("raftexample: create wal error: %v", err)
+		}
+		w.Close()
+	}
+
+	w, err := wal.Open(waldir, walpb.Snapshot{})
+	if err != nil {
+		log.Fatalf("raftexample: error loading wal (%v)", err)
+	}
+
+	return w
+}
+
+// Replay commits already in WAL over commit channel. Returns an appendable WAL.
+func (rc *raftKVChan) replayWAL() *wal.WAL {
+	w := rc.openWAL()
+	_, _, ents, err := w.ReadAll()
+	if err != nil {
+		log.Fatalf("raftexample: failed to read WAL (%v)", err)
+	}
+	// append to storage so raft starts at the right place in log
+	rc.raftStorage.Append(ents)
+	rc.publishEntries(ents)
+	// send nil value so client knows commit channel is current
+	rc.commitC <- raftKV{Val : nil}
+	return w
+}
+
+func (rc *raftKVChan) writeError(err error) {
+	rc.errorC <- err
+	rc.stop()
+}
+
+func (rc *raftKVChan) stop() {
+	close(rc.commitC)
+	close(rc.errorC)
+	rc.node.Stop()
+}
+
+func (rc *raftKVChan) startRaft() {
+	rc.wal = rc.replayWAL()
+
+	rpeers := make([]raft.Peer, len(rc.peers))
+	for i := range rpeers {
+		rpeers[i] = raft.Peer{ID : uint64(i+1)}
+	}
+	c := &raft.Config{
+		ID:              uint64(rc.id),
+		ElectionTick:    10,
+		HeartbeatTick:   1,
+		Storage:         rc.raftStorage,
+		MaxSizePerMsg:   1024 * 1024,
+		MaxInflightMsgs: 256,
+	}
+	rc.node = raft.StartNode(c, rpeers)
+
+	ss := &stats.ServerStats{}
+	ss.Initialize()
+
+	rc.transport = &rafthttp.Transport{
+		ID : types.ID(rc.id),
+		ClusterID : 0x1000,
+		Raft : rc,
+		ServerStats : ss,
+		LeaderStats : stats.NewLeaderStats(strconv.Itoa(rc.id)),
+		ErrorC : make(chan error),
+	}
+
+	rc.transport.Start()
+	for i := range rc.peers {
+		if i+1 != rc.id {
+			rc.transport.AddPeer(types.ID(i+1), []string{rc.peers[i]})
+		}
+	}
+
+	go rc.serveRaft()
+	go rc.serveChannels()
+}
+
+func (rc *raftKVChan) serveChannels() {
+	defer rc.wal.Close()
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	// event loop on client proposals and raft updates
+	for {
+		select {
+		case <-ticker.C:
+			rc.node.Tick()
+
+		// send proposals over raft
+		case prop, ok := <-rc.proposeC:
+			if !ok {
+				// client closed channel; shut down
+				rc.stop()
+				return
+			}
+			req := etcdserverpb.Request{
+				Method : "PUT",
+				Path : prop.Key,
+				Val : string(prop.Val),}
+			data, err := req.Marshal()
+			if err != nil {
+				rc.writeError(err)
+				return
+			}
+			rc.node.Propose(context.TODO(), data)
+
+		// store raft entries to wal, then publish over commit channel
+		case rd := <-rc.node.Ready():
+			rc.wal.Save(rd.HardState, rd.Entries)
+			rc.raftStorage.Append(rd.Entries)
+			rc.transport.Send(rd.Messages)
+			rc.publishEntries(rd.Entries)
+			rc.node.Advance()
+
+		case err := <-rc.transport.ErrorC:
+			rc.writeError(err)
+			return
+		}
+	}
+}
+
+func (rc *raftKVChan) serveRaft() {
+	// strip off http:// so peer address can be used for server
+	addr := strings.Split(rc.peers[rc.id-1], "/")[2]
```

Comment:
url.Parse will do the work for you and make this more readable in my opinion. Or if we only support http we can probably just ask user to give ip:port without http scheme.